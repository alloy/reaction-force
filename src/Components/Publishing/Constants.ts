import cheerio from "cheerio"
import { compact, last, uniq } from "lodash"
import moment from "moment"
import momentTz from "moment-timezone"
import url from "url"
import { ArticleData, DateFormat } from "../Publishing/Typings"

const APP_URL = process.env.APP_URL || "https://www.artsy.net"

/**
 * Matches for Email / Instant Articles
 */
export const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

/**
 * The quality to request from image CDN
 */
export const GLOBAL_IMAGE_QUALITY = 80

/**
 * TODO: Eventually remove sizeMe
 */
export const SIZE_ME_REFRESH_RATE = 500

/**
 * Relative path to article
 */
export const getArticleHref = slug => `/article/${slug}`

/**
 * Absolute path to article
 */
export const getArticleFullHref = slug => `${APP_URL}/article/${slug}`

/**
 * Get the pre-slug part of URL
 */

export const getPreSlugPath = layout => {
  return ["standard", "feature"].includes(layout) ? "article" : layout
}

/**
 * Relative path to editorial entity
 */

export const getEditorialHref = (layout, slug) => {
  const layoutType = getPreSlugPath(layout)
  return `/${layoutType}/${slug}`
}

/**
 * Absolute path to editorial entity
 */
export const getFullEditorialHref = (layout, slug) => {
  const layoutType = getPreSlugPath(layout)
  return `${APP_URL}/${layoutType}/${slug}`
}

/**
 * Absolute path to artsy entity
 */
export const getFullArtsyHref = slug => `${APP_URL}/${slug}`

/**
 * ByLine helpers
 * TODO: Move this into some kind of utils folder
 */
export const getAuthorByline = (authors, isEditoral = true) => {
  const authorCount = Number(authors && authors.length)

  if (authorCount === 1) {
    return authors[0].name || ""
  } else if (authorCount > 1) {
    const names = authors.reduce((prev, curr, i) => {
      let delim
      const len = authors.length
      if (i === len - 1) {
        delim = " and "
      } else if (i === 0) {
        delim = ""
      } else {
        delim = ", "
      }
      return prev + delim + curr.name
    }, "")
    return names

    // No Author
  } else if (isEditoral) {
    return "Artsy Editors"
  }
}

export const getDate = (date, format: DateFormat = "default") => {
  const today = moment()
  const isToday = today.isSame(moment(date), "day")
  const isThisYear =
    moment(date).format("YYYY") === moment(today).format("YYYY")

  switch (format) {
    case "monthDay":
      return moment(date).format("MMM D")
    case "monthYear":
      return momentTz(date)
        .tz("America/New_York")
        .format("MMMM YYYY")
    case "condensed":
      return momentTz(date)
        .tz("America/New_York")
        .format("MMM D, YYYY")
    case "verbose":
      const day = isToday
        ? "Today"
        : momentTz(date)
            .tz("America/New_York")
            .format("MMM D, YYYY")
      const time = momentTz(date)
        .tz("America/New_York")
        .format("h:mm a")
      return `${day} at ${time}`
    case "news":
      return isToday
        ? "Today"
        : isThisYear
          ? momentTz(date)
              .tz("America/New_York")
              .format("MMM D")
          : momentTz(date)
              .tz("America/New_York")
              .format("MMM D, YYYY")
    default:
      return momentTz(date)
        .tz("America/New_York")
        .format("MMM D, YYYY h:mm a")
  }
}

export const getCurrentUnixTimestamp = () => moment().unix()

export const getMediaDate = article => {
  const { published_at, scheduled_publish_at, media } = article
  const { release_date } = media

  if (release_date) {
    return release_date
  } else {
    return published_at || scheduled_publish_at
  }
}

export const formatTime = time => {
  let minutes = Math.floor(time / 60) % 60
  let seconds = Math.floor(time % 60)
  minutes = minutes <= 0 ? 0 : minutes
  seconds = seconds <= 0 ? 0 : seconds

  const minutesStr = minutes < 10 ? "0" + minutes : minutes
  const secondsStr = seconds < 10 ? "0" + seconds : seconds
  return minutesStr + ":" + secondsStr
}

interface SlugsFromArticle {
  artists: string[]
  genes: string[]
}

export const getArtsySlugsFromArticle = (
  article: ArticleData
): SlugsFromArticle => {
  const articleBody = article.sections
    .map(section => {
      if (section.type === "text") {
        return section.body
      }
    })
    .join()

  const artists = uniq(getArtsySlugsFromHTML(articleBody, "artist"))
  const genes = uniq(getArtsySlugsFromHTML(articleBody, "gene"))

  return {
    artists,
    genes,
  }
}

export const getArtsySlugsFromHTML = (
  html: string,
  model: string
): string[] => {
  const $ = cheerio.load(html)

  const slugs = compact($("a")).map(a => {
    const href = $(a).attr("href")
    if (href) {
      if (href.match(`artsy.net/${model}`)) {
        return last(url.parse(href).pathname.split("/"))
      } else {
        return null
      }
    } else {
      return null
    }
  })
  return compact(slugs)
}
