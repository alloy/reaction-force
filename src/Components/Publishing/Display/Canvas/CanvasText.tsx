import { garamond } from "Assets/Fonts"
import { pMedia } from "Components/Helpers"
import { compact, map } from "lodash"
import React from "react"
import styled, { StyledFunction } from "styled-components"
import { resize } from "Utils/resizer"

interface CanvasTextProps {
  disclaimer?: any
  unit: any
}

interface DivProps extends React.HTMLProps<HTMLDivElement> {
  layout?: any
  isSlideshowWithCaption?: boolean
}

export const CanvasText: React.SFC<CanvasTextProps> = props => {
  const { disclaimer, unit } = props
  const resized_logo = unit.logo
    ? resize(unit.logo, {
        // slightly oversize to avoid fuzziness
        width: 400,
        isDisplayAd: true,
      })
    : ""
  const isSlideshowWithCaption =
    unit.layout === "slideshow" && compact(map(unit.assets, "caption")).length
  const hasDisclaimer = unit.layout !== "overflow" && !isSlideshowWithCaption
  return (
    <CanvasInner layout={unit.layout} isSlideshowWithCaption>
      <div>
        <Logo layout={unit.layout} src={resized_logo} />
        <Headline layout={unit.layout}>{unit.headline}</Headline>
        <Link layout={unit.layout}>{unit.link.text}</Link>
      </div>
      {hasDisclaimer && disclaimer}
    </CanvasInner>
  )
}

const Div: StyledFunction<DivProps> = styled.div
const Img: StyledFunction<DivProps> = styled.img

const Logo = Img`
  height: 100%;
  object-fit: contain;
  object-position: left;
  max-width: ${props => (props.layout === "overlay" ? "300px;" : "250px;")}
  max-height: ${props => (props.layout === "overlay" ? "100px;" : "90px;")}
  margin: ${props =>
    props.layout === "overlay" ? "60px auto;" : "20px 0 50px 0;"}
  ${props => pMedia.md`
    width: auto;
    margin: 20px 0;
    ${props.layout === "overlay" &&
      `max-width: calc(100% - 60px);
       margin: 15px auto;
    `}
  `}
`
const CanvasInner = Div`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: ${props =>
    props.layout === "overlay" || props.isSlideshowWithCaption
      ? "space-around;"
      : "space-between;"}
  ${props =>
    props.layout === "overlay" &&
    `text-align: center;
        color: #fff;
        max-width: 500px;`}
  ${props => pMedia.md`
    ${props.layout === "slideshow" && "width: 100%;"}
  `}
  ${pMedia.sm`
    max-width: 100%;
    width: 100%;
    padding: 0;
  `}
`
const Headline = Div`
  ${props => (props.layout === "overlay" ? garamond("s23") : garamond("s40"))}
  line-height: ${props => (props.layout === "overlay" ? "1.35em;" : "1.1em;")} 
  margin-bottom: 25px;
  ${props => pMedia.lg`
    ${props.layout !== "overlay" && garamond("s23")}
  `}
  ${props => pMedia.sm`
    margin: 0 auto 10px auto;
    max-width: calc(100% - 40px);
    ${props.layout !== "overlay" && "margin-left: 0;"}
  `}
  ${props => pMedia.xs`
    ${props.layout === "overlay" && garamond("s17")}
    line-height: 1.35em;
  `}
`
const Link = Div`
  ${garamond("s23")}
  line-height: 1.35em;
  margin-bottom: 10px;
  display: initial;
  background-image: linear-gradient(to bottom, transparent 0, ${props =>
    props.layout === "overlay" ? "#FFF" : "#000"} 2px, transparent 0);
  background-position: bottom;
  background-size: 1px 5px;
  background-repeat: repeat-x;
  transition: all .3s;
  &:hover {
    opacity: .6;
  }
  ${props => pMedia.lg`
    ${props.layout !== "overlay" && garamond("s17")}
  `}
  ${pMedia.md`
    ${garamond("s17")}
    line-height: 1.35em;
  `}
`
