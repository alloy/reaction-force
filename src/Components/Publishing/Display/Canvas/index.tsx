import React from "react"
import styled, { StyledFunction } from "styled-components"
import Colors from "../../../../Assets/Colors"
import { pMedia } from "../../../Helpers"
import { Fonts } from "../../Fonts"
import { CanvasContainer } from "./CanvasContainer"

interface DisplayCanvasProps {
  unit: any
  campaign: any
}

export const DisplayCanvas: React.SFC<DisplayCanvasProps> = props => {
  const { unit, campaign } = props
  const disclaimer = <Disclaimer layout={unit.layout}>{unit.disclaimer}</Disclaimer>

  return (
    <DisplayContainer layout={unit.layout}>
      <a href={unit.link.url} target="_blank">
        <AdvertisementBy>{`Advertisement by ${campaign.name}`}</AdvertisementBy>
      </a>
      <CanvasContainer unit={unit} campaign={campaign} disclaimer={disclaimer} />
      {unit.layout === "overlay" && disclaimer}
    </DisplayContainer>
  )
}

interface DivProps extends React.HTMLProps<HTMLDivElement> {
  layout: string
}

const Div: StyledFunction<DivProps> = styled.div

const DisplayContainer = Div`
  display: flex;
  flex-direction: column;
  width: 100%;
  max-width: ${props => (props.layout === "slideshow" ? "100%;" : "1250px;")}
  margin: 0 auto;
  margin-bottom: -44px; // Offset default force margin; FIXME: This kind of stuff needs to be globally addressed
  box-sizing: border-box;
  a {
    text-decoration: none;
  }
  ${props => pMedia.sm`
    ${props.layout !== "slideshow" && "padding: 0 20px;"}
    margin-bottom: 0;
  `}
`
const AdvertisementBy = styled.div`
  ${Fonts.avantgarde("s11")}
  color: ${Colors.grayMedium};
  margin: 10px 0;
  text-align: center;
`
const Disclaimer = Div`
  ${Fonts.garamond("s11")}
  color: ${Colors.grayMedium};
  margin: 15px 0 0 0;
  ${props => props.layout === "overlay" && "text-align: center;"}
  ${pMedia.sm`
    margin: 35px 0 10px 0;
  `}
`