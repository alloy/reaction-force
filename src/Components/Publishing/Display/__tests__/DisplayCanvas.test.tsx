import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import { getCurrentUnixTimestamp } from "../../Constants"
import { DisplayCanvas } from "../Canvas"
import { CanvasLink } from "../Canvas/CanvasContainer"
import { CanvasSlideshow } from "../Canvas/CanvasSlideshow"
import { CanvasText } from "../Canvas/CanvasText"
import { CanvasVideo } from "../Canvas/CanvasVideo"

import {
  Campaign,
  HostedAdCanvas,
  UnitCanvasImage,
  UnitCanvasOverlay,
  UnitCanvasSlideshow,
  UnitCanvasTracked,
  UnitCanvasVideo,
} from "../../Fixtures/Components"

describe("snapshot", () => {
  xit("renders the canvas in standard layout with image", () => {
    const component = renderer
      .create(
        <DisplayCanvas
          unit={UnitCanvasImage}
          campaign={Campaign}
          adUnit={HostedAdCanvas.adUnit}
          adDimension={HostedAdCanvas.adDimension}
        />
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })

  it("renders the canvas in standard layout with video", () => {
    const component = renderer
      .create(
        <DisplayCanvas
          unit={UnitCanvasVideo}
          campaign={Campaign}
          adUnit={HostedAdCanvas.adUnit}
          adDimension={HostedAdCanvas.adDimension}
        />
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })

  it("renders the canvas in overlay layout", () => {
    const component = renderer
      .create(
        <DisplayCanvas
          unit={UnitCanvasOverlay}
          campaign={Campaign}
          adUnit={HostedAdCanvas.adUnit}
          adDimension={HostedAdCanvas.adDimension}
        />
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })

  it("renders the canvas in slideshow layout", () => {
    const component = renderer
      .create(
        <DisplayCanvas
          unit={UnitCanvasSlideshow}
          campaign={Campaign}
          adUnit={HostedAdCanvas.adUnit}
          adDimension={HostedAdCanvas.adDimension}
        />
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })
})

describe("unit", () => {
  const trackEvent = jest.fn()
  const getWrapper = (props: any = {}) => {
    return mount(
      <DisplayCanvas
        unit={props.unit || UnitCanvasImage}
        campaign={props.campaign || Campaign}
        tracking={{
          trackEvent,
        }}
        renderTime={12345}
        adUnit={HostedAdCanvas.adUnit}
        adDimension={HostedAdCanvas.adDimension}
      />
    )
  }

  xit("renders the unit data", () => {
    const canvas = mount(
      <DisplayCanvas
        unit={UnitCanvasImage}
        campaign={Campaign}
        adUnit={HostedAdCanvas.adUnit}
        adDimension={HostedAdCanvas.adDimension}
      />
    )
    expect(canvas.html()).toMatch(UnitCanvasImage.disclaimer)
    expect(canvas.html()).toMatch(UnitCanvasImage.headline)
    expect(canvas.html()).toMatch(UnitCanvasImage.link.text)
    expect(canvas.html()).toMatch(UnitCanvasImage.link.url)
    expect(canvas.html()).toMatch("artsy-logo-wide-black")
    expect(canvas.html()).toMatch("cloudfront")
    expect(canvas.find(CanvasText).length).toBe(1)
    expect(canvas.find(CanvasSlideshow).length).toBe(0)
    expect(canvas.find(CanvasVideo).length).toBe(0)
  })

  it("renders the canvas with the correct dimensions and unit name", () => {
    const canvas = mount(
      <DisplayCanvas
        unit={UnitCanvasImage}
        campaign={Campaign}
        adUnit={HostedAdCanvas.adUnit}
        adDimension={HostedAdCanvas.adDimension}
      />
    )

    expect(canvas.html()).toMatch(HostedAdCanvas.adUnit)
    expect(canvas.html()).toMatch(HostedAdCanvas.adDimension)
  })

  xit("renders the video component if standard layout with video", () => {
    const canvas = mount(
      <DisplayCanvas
        unit={UnitCanvasVideo}
        campaign={Campaign}
        adUnit={HostedAdCanvas.adUnit}
        adDimension={HostedAdCanvas.adDimension}
      />
    )
    expect(canvas.find(CanvasVideo).length).toBe(1)
  })

  xit("renders the slideshow component if slideshow layout", () => {
    const canvas = mount(
      <DisplayCanvas
        unit={UnitCanvasSlideshow}
        campaign={Campaign}
        adUnit={HostedAdCanvas.adUnit}
        adDimension={HostedAdCanvas.adDimension}
      />
    )
    expect(canvas.find(CanvasSlideshow).length).toBe(1)
  })

  xit("renders a pixel impression if there is a url", () => {
    const wrapper = getWrapper({
      unit: UnitCanvasTracked,
    })
    expect(wrapper.html()).toMatch("impression?ord=12345")
  })

  describe("analytics", () => {
    it("tracks impressions", () => {
      const wrapper = getWrapper() as any
      wrapper.instance().trackImpression()

      expect(trackEvent.mock.calls[0][0]).toEqual(
        expect.objectContaining({
          action: "Impression",
          entity_type: "display_ad",
          campaign_name: "Artsy",
          unit_layout: "canvas_standard",
        })
      )
    })

    it("tracks viewability", () => {
      const wrapper = getWrapper() as any
      wrapper.instance().trackViewability()

      expect(trackEvent.mock.calls[1][0]).toEqual(
        expect.objectContaining({
          action: "Viewability",
          entity_type: "display_ad",
          campaign_name: "Artsy",
          unit_layout: "canvas_standard",
        })
      )
    })

    xit("busts 3rd party cache on click", () => {
      const wrapper = getWrapper({
        unit: UnitCanvasTracked,
      })
      const currentTime = getCurrentUnixTimestamp()
        .toString()
        .substring(6)
      wrapper.find(CanvasLink).simulate("click")
      const open = window.open as any
      expect(open.mock.calls[0][0]).toMatch("?ord=")
      expect(open.mock.calls[0][0]).toMatch(currentTime)
    })
  })
})
