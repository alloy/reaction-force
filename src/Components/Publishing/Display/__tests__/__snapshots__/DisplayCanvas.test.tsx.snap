// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot renders the canvas in overlay layout 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  max-width: 1250px;
  margin: 0 auto;
  box-sizing: border-box;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <div
    className="htl-ad"
    data-eager={true}
    data-sizes="970x250"
    data-unit="Desktop_TopLeaderboard"
  />
</div>
`;

exports[`snapshot renders the canvas in slideshow layout 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  max-width: 100%;
  margin: 0 auto;
  box-sizing: border-box;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <div
    className="htl-ad"
    data-eager={true}
    data-sizes="970x250"
    data-unit="Desktop_TopLeaderboard"
  />
</div>
`;

exports[`snapshot renders the canvas in standard layout with image 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  max-width: 1250px;
  margin: 0 auto;
  box-sizing: border-box;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <div
    className="htl-ad"
    data-eager={true}
    data-sizes="970x250"
    data-unit="Desktop_TopLeaderboard"
  />
</div>
`;

exports[`snapshot renders the canvas in standard layout with video 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  max-width: 1250px;
  margin: 0 auto;
  box-sizing: border-box;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="c0"
>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <div
    className="htl-ad"
    data-eager={true}
    data-sizes="970x250"
    data-unit="Desktop_TopLeaderboard"
  />
</div>
`;
